client/src/client.ts

```diff
@@ -66,9 +66,7 @@ export class AngularLanguageClient implements vscode.Disposable {
         provideCodeActions: async (
             document: vscode.TextDocument, range: vscode.Range, context: vscode.CodeActionContext,
             token: vscode.CancellationToken, next: lsp.ProvideCodeActionsSignature) => {
-          if (await this.isInAngularProject(document) &&
-              isInsideInlineTemplateRegion(document, range.start) &&
-              isInsideInlineTemplateRegion(document, range.end)) {
+          if (await this.isInAngularProject(document)) {
```

Comment:
Hmm, we should keep an eye on reports of performance regressions with this change. I don't know if this particular request depends greatly on quick turnaround like completions would. 

`onCodeAction` does call `getCodeFixesAtPosition` which triggers semantic diagnostics (and I think global compilation) so it could be pretty slow in a large project/file. Maybe we could optimize this by adding an endpoint to the language service that can first check if there's a `CodeActionMeta` for the diagnostic error code and only then ask for the code fixes.

Reply:
I think this is reasonable. Let me look into this.  

Though I wonder if the performance problem you were addressing back then, actually was addressed by this. Based on the issue conversation it didn't seem like it?

Reply:
When I call the `getSemanticDiagnostics` in the code action, I think it should not have a performance issue.

1. This only triggers the currently active file semantic diagnostics. The time spent should be the same as the completion.

2. The vscode should trigger the code action after the diagnostics changes. So When the code action is triggered, the semantic diagnostics have been computed. the second time triggering the diagnostics should hit the cache(If not, I think the code action can be canceled, and wait for the vscode to trigger the semantic diagnostics first). We can cache the diagnostics in the `program`(or `SourceFile`) or the `NgCompiler` has done it?

Correct me if I'm wrong.

Reply:
I'm not too familiar with this code in concrete here, but couldn't VSCode trigger a code action on a file that only has semantic diagnostics checked for TypeScript— but not the Angular extension?

I assume, probably not the case, but in either case, it wouldn't hurt having a safety quick exit just to be sure?

Reply:
> This only triggers the currently active file semantic diagnostics. The time spent should be the same as the completion.

Yes, but this actually still requires the whole program to be analyzed:

https://github.com/angular/angular/blob/982f1b125147e4292716f9524bef75423b70c71c/packages/compiler-cli/src/ngtsc/core/src/compiler.ts#L596
https://github.com/angular/angular/blob/982f1b125147e4292716f9524bef75423b70c71c/packages/compiler-cli/src/ngtsc/core/src/compiler.ts#L1156
https://github.com/angular/angular/blob/982f1b125147e4292716f9524bef75423b70c71c/packages/compiler-cli/src/ngtsc/core/src/compiler.ts#L930-L951

Reply:
> I assume, probably not the case, and what you write makes sense to me— but in either case, it wouldn't hurt having a safety quick exit just to be sure?

No, it wouldn't hurt to have this. I am just curious about what caused this performance issue.

> I'm not too familiar with this code in concrete here, but couldn't VSCode trigger a code action on a file that only has semantic diagnostics checked for TypeScript— but not the Angular extension?

The code action will be skipped because it's out of the angular context.

> Yes, but this actually still requires the whole program to be analyzed:

But the completion also requires the whole program to be analyzed, because it depends on the template type checker:

https://github.com/angular/angular/blob/d97aee6dc2266c81cbd5207f642f3fee39a2d414/packages/language-service/src/completions.ts#L124

https://github.com/angular/angular/blob/d97aee6dc2266c81cbd5207f642f3fee39a2d414/packages/compiler-cli/src/ngtsc/core/src/compiler.ts#L701

Reply:
> But the completion also requires the whole program to be analyzed, because it depends on the template type checker

Yes! Pretty much everything requires it and that’s why we try to avoid making these requests in places where we don’t have to. When we didn’t have these early-returns, we had performance issues files in the repo, especially for completions, where we were slowing down the completion results for every typescript file, even outside the inline template where we’d have no additional completion information.

Reply:
Thanks, it makes sense to me.